### PR TITLE
use RecordBuilders in CustomDimensions plugin

### PR DESCRIPTION
**Pull request by diosmosis**
### Description:

**This PR is based off of #20719 and https://github.com/matomo-org/matomo/pull/20684 which should be reviewed and merged first.**

Changes:
* Move archiving code in CustomDimensions to RecordBuilder that aggregates for a single dimension.
* Create CustomDimensions RecordBuilders for each custom dimension via an event.
* Remove now unused DataArray instance.
* Add filter EnrichRecordWithGoalMetricSums to replace DataArray::enrichMetricsWithConversions().
* Allow descendants of RecordBuilders to manually insert tables if needed (will be used by Cohorts).

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: core/Metrics.php [hunk 1]
```diff
@@ -581,4 +582,39 @@ public static function getPercentVisitColumn()
         $percentVisitsLabel = str_replace(' ', '&nbsp;', Piwik::translate('General_ColumnPercentageVisits'));
         return $percentVisitsLabel;
     }
+
+    public static function makeGoalColumnsRow(int $idGoal, array $goals): array
```
Comment (sgiehl) [inline core/Metrics.php]: This one could have a comment what it should/can be used for.

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+            foreach ($row as $name => $columnValue) {
+                $columns[$name] = $columnValue;
+            }
+
+            $report->sumRowWithLabel($customDimensionValue, $columns);
+        }
+    }
+
+    private function aggregateFromConversions(DataTable $report, LogAggregator $logAggregator, string $valueField,
+                                              array $dimensions, string $where)
```
Comment (sgiehl) [inline plugins/CustomDimensions/RecordBuilders/CustomDimension.php]:
```suggestion
                                              array $dimensions, string $where): void
```

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                Metrics::INDEX_GOALS => [
+                    $idGoal => Metrics::makeGoalColumnsRow($idGoal, $row),
+                ],
+            ];
+
+            $report->sumRowWithLabel($value, $columns);
+        }
+    }
+
+    protected function aggregateFromActions(DataTable $report, LogAggregator $logAggregator, $valueField)
```
Comment (sgiehl) [inline plugins/CustomDimensions/RecordBuilders/CustomDimension.php]:
```suggestion
    protected function aggregateFromActions(DataTable $report, LogAggregator $logAggregator, $valueField): void
```

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    protected function cleanCustomDimensionValue(string $value): string
+    {
+        if (isset($value) && strlen($value)) {
+            return $value;
+        }
+
+        return Archiver::LABEL_CUSTOM_VALUE_NOT_DEFINED;
+    }
+
+    private function addMetricsToSelect($select, $metricsConfig)
```
Comment (sgiehl) [inline plugins/CustomDimensions/RecordBuilders/CustomDimension.php]:
```suggestion
    private function addMetricsToSelect(string $select, array $metricsConfig): string
```

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            "log_link_visit_action",
+            array(
+                "table"  => "log_visit",
+                "joinOn" => "log_visit.idvisit = log_link_visit_action.idvisit"
+            ),
+            array(
+                "table"  => "log_action",
+                "joinOn" => "log_link_visit_action.idaction_url = log_action.idaction"
+            )
+        );
```
Comment (sgiehl) [inline plugins/CustomDimensions/RecordBuilders/CustomDimension.php]:
```suggestion
        $from = [
            "log_link_visit_action",
            [
                "table"  => "log_visit",
                "joinOn" => "log_visit.idvisit = log_link_visit_action.idvisit"
            ],
            [
                "table"  => "log_action",
                "joinOn" => "log_link_visit_action.idaction_url = log_action.idaction"
            ]
        ];
```

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+        $groupBy = "log_link_visit_action.$valueField, url";
+        $orderBy = "`" . Metrics::INDEX_PAGE_NB_HITS . "` DESC";
+
+        // get query with segmentation
+        $query     = $logAggregator->generateQuery($select, $from, $where, $groupBy, $orderBy);
+
+        if ($this->rankingQueryLimit > 0) {
+            $rankingQuery = new RankingQuery($this->rankingQueryLimit);
+            $rankingQuery->addLabelColumn(array($valueField, 'url'));
```
Comment (sgiehl) [inline plugins/CustomDimensions/RecordBuilders/CustomDimension.php]:
```suggestion
            $rankingQuery->addLabelColumn([$valueField, 'url']);
```

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            return [];
+        }
+
+        $logAggregator = $archiveProcessor->getLogAggregator();
+
+        $report = new DataTable();
+        $recordName = Archiver::buildRecordNameForCustomDimensionId($dimension['idcustomdimension']);
+
+        $valueField = LogTable::buildCustomDimensionColumnName($dimension);
+        $dimensions = array($valueField);
```
Comment (sgiehl) [inline plugins/CustomDimensions/RecordBuilders/CustomDimension.php]:
```suggestion
        $dimensions = [$valueField];
```

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+            $rankingQuery = new RankingQuery($this->rankingQueryLimit);
+            $rankingQuery->addLabelColumn($dimensions[0]);
+
+            $query = $logAggregator->queryVisitsByDimension($dimensions, $where, [], false, $rankingQuery, false, -1,
+                $rankingQueryGenerate = true);
+        } else {
+            $query = $logAggregator->queryVisitsByDimension($dimensions, $where);
+        }
+
+        $defaultColumns = DataArray::makeEmptyRow();
```
Comment (sgiehl) [inline plugins/CustomDimensions/RecordBuilders/CustomDimension.php]: Aren't we tryng to get rid of the DataArray stuff? Wondering if there's already another method that could be used instead.

File: core/DataTable/Filter/EnrichRecordWithGoalMetricSums.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            $row->setColumns($columns);
+
+            $subtable = $row->getSubtable();
+            if ($subtable) {
+                $this->filter($subtable);
+            }
+        }
+    }
+
+    public static function enrichWithConversions(&$values)
```
Comment (sgiehl) [inline core/DataTable/Filter/EnrichRecordWithGoalMetricSums.php]:
```suggestion
    public static function enrichWithConversions(&$values): void
```

**Review by sgiehl — CHANGES_REQUESTED** [review]
Overall looks fine. Left a couple of comments for improvements.

File: plugins/CustomDimensions/RecordBuilders/CustomDimension.php (hunk 8 above)
Comment (diosmosis) [reply plugins/CustomDimensions/RecordBuilders/CustomDimension.php]: Currently there isn't another method. There might be if the simpler log aggregation poc is accepted. That won't be discussed for a while though. Replaced with just setting the column values to 0 directly.

**Review by sgiehl — APPROVED** [review]

**Pull request merged**